In [60]:
import pandas as pd

# Load and Inspect Raw Data

In [61]:
orders = pd.read_csv(
    "../01_data/raw/olist_orders_dataset.csv"
)

items = pd.read_csv(
    "../01_data/raw/olist_order_items_dataset.csv"
)

products = pd.read_csv(
    "../01_data/raw/olist_products_dataset.csv"
)

sellers = pd.read_csv(
    "../01_data/raw/olist_sellers_dataset.csv"
)

customers = pd.read_csv(
    "../01_data/raw/olist_customers_dataset.csv"
)

payments = pd.read_csv(
    "../01_data/raw/olist_order_payments_dataset.csv"
)

reviews = pd.read_csv(
    "../01_data/raw/olist_order_reviews_dataset.csv"
)

for name, df in {
    "orders": orders,
    "items": items,
    "products": products,
    "sellers": sellers,
    "customers": customers,
    "payments": payments,
    "reviews": reviews,
}.items():
    print(f"\n{name.upper()}")
    print(df.shape)
    display(df.info())


ORDERS
(99441, 8)
<class 'pandas.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype
---  ------                         --------------  -----
 0   order_id                       99441 non-null  str  
 1   customer_id                    99441 non-null  str  
 2   order_status                   99441 non-null  str  
 3   order_purchase_timestamp       99441 non-null  str  
 4   order_approved_at              99281 non-null  str  
 5   order_delivered_carrier_date   97658 non-null  str  
 6   order_delivered_customer_date  96476 non-null  str  
 7   order_estimated_delivery_date  99441 non-null  str  
dtypes: str(8)
memory usage: 6.1 MB


None


ITEMS
(112650, 7)
<class 'pandas.DataFrame'>
RangeIndex: 112650 entries, 0 to 112649
Data columns (total 7 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   order_id             112650 non-null  str    
 1   order_item_id        112650 non-null  int64  
 2   product_id           112650 non-null  str    
 3   seller_id            112650 non-null  str    
 4   shipping_limit_date  112650 non-null  str    
 5   price                112650 non-null  float64
 6   freight_value        112650 non-null  float64
dtypes: float64(2), int64(1), str(4)
memory usage: 6.0 MB


None


PRODUCTS
(32951, 9)
<class 'pandas.DataFrame'>
RangeIndex: 32951 entries, 0 to 32950
Data columns (total 9 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   product_id                  32951 non-null  str    
 1   product_category_name       32341 non-null  str    
 2   product_name_lenght         32341 non-null  float64
 3   product_description_lenght  32341 non-null  float64
 4   product_photos_qty          32341 non-null  float64
 5   product_weight_g            32949 non-null  float64
 6   product_length_cm           32949 non-null  float64
 7   product_height_cm           32949 non-null  float64
 8   product_width_cm            32949 non-null  float64
dtypes: float64(7), str(2)
memory usage: 2.3 MB


None


SELLERS
(3095, 4)
<class 'pandas.DataFrame'>
RangeIndex: 3095 entries, 0 to 3094
Data columns (total 4 columns):
 #   Column                  Non-Null Count  Dtype
---  ------                  --------------  -----
 0   seller_id               3095 non-null   str  
 1   seller_zip_code_prefix  3095 non-null   int64
 2   seller_city             3095 non-null   str  
 3   seller_state            3095 non-null   str  
dtypes: int64(1), str(3)
memory usage: 96.8 KB


None


CUSTOMERS
(99441, 5)
<class 'pandas.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 5 columns):
 #   Column                    Non-Null Count  Dtype
---  ------                    --------------  -----
 0   customer_id               99441 non-null  str  
 1   customer_unique_id        99441 non-null  str  
 2   customer_zip_code_prefix  99441 non-null  int64
 3   customer_city             99441 non-null  str  
 4   customer_state            99441 non-null  str  
dtypes: int64(1), str(4)
memory usage: 3.8 MB


None


PAYMENTS
(103886, 5)
<class 'pandas.DataFrame'>
RangeIndex: 103886 entries, 0 to 103885
Data columns (total 5 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   order_id              103886 non-null  str    
 1   payment_sequential    103886 non-null  int64  
 2   payment_type          103886 non-null  str    
 3   payment_installments  103886 non-null  int64  
 4   payment_value         103886 non-null  float64
dtypes: float64(1), int64(2), str(2)
memory usage: 4.0 MB


None


REVIEWS
(99224, 7)
<class 'pandas.DataFrame'>
RangeIndex: 99224 entries, 0 to 99223
Data columns (total 7 columns):
 #   Column                   Non-Null Count  Dtype
---  ------                   --------------  -----
 0   review_id                99224 non-null  str  
 1   order_id                 99224 non-null  str  
 2   review_score             99224 non-null  int64
 3   review_comment_title     11568 non-null  str  
 4   review_comment_message   40977 non-null  str  
 5   review_creation_date     99224 non-null  str  
 6   review_answer_timestamp  99224 non-null  str  
dtypes: int64(1), str(6)
memory usage: 5.3 MB


None

In [62]:
print("Orders")
print("Rows:", len(orders))
print("Unique order IDs:", orders["order_id"].nunique())
print("Unique customer IDs:", orders["customer_id"].nunique())

print("\nItems")
print("Rows:", len(items))
print("Unique order IDs:", items["order_id"].nunique())
print("Unique product IDs:", items["product_id"].nunique())
print("Unique seller IDs:", items["seller_id"].nunique())

print("\nProducts")
print("Rows:", len(products))
print("Unique product IDs:", products["product_id"].nunique())

print("\nSellers")
print("Rows:", len(sellers))
print("Unique seller IDs:", sellers["seller_id"].nunique())

print("\nCustomers")
print("Rows:", len(customers))
print("Unique customer IDs:", customers["customer_id"].nunique())
print("Unique customer unique IDs:", customers["customer_unique_id"].nunique())

print("\nPayments")
print("Rows:", len(payments))
print("Unique order IDs:", payments["order_id"].nunique())

print("\nReviews")
print("Rows:", len(reviews))
print("Unique review IDs:", reviews["review_id"].nunique())
print("Unique order IDs:", reviews["order_id"].nunique())

Orders
Rows: 99441
Unique order IDs: 99441
Unique customer IDs: 99441

Items
Rows: 112650
Unique order IDs: 98666
Unique product IDs: 32951
Unique seller IDs: 3095

Products
Rows: 32951
Unique product IDs: 32951

Sellers
Rows: 3095
Unique seller IDs: 3095

Customers
Rows: 99441
Unique customer IDs: 99441
Unique customer unique IDs: 96096

Payments
Rows: 103886
Unique order IDs: 99440

Reviews
Rows: 99224
Unique review IDs: 98410
Unique order IDs: 98673


**Customer identifiers:** `customer_id` identifies the customer record associated
with an individual order, while `customer_unique_id` provides a persistent
identifier for the same underlying customer across multiple orders. Therefore,
`customer_unique_id` should be used when analyzing repeat purchasing or
customer-level behavior across orders.

In [63]:
items_per_order = items.groupby("order_id").size()
sellers_per_order = items.groupby("order_id")["seller_id"].nunique()
products_per_order = items.groupby("order_id")["product_id"].nunique()
payments_per_order = payments.groupby("order_id").size()
reviews_per_order = reviews.groupby("order_id").size()

print("Items per order")
display(items_per_order.describe())

print("\nSellers per order")
display(sellers_per_order.value_counts().sort_index())

print("\nProducts per order")
display(products_per_order.describe())

print("\nPayments per order")
display(payments_per_order.value_counts().sort_index())

print("\nReviews per order")
display(reviews_per_order.value_counts().sort_index())

Items per order


count    98666.000000
mean         1.141731
std          0.538452
min          1.000000
25%          1.000000
50%          1.000000
75%          1.000000
max         21.000000
dtype: float64


Sellers per order


seller_id
1    97388
2     1219
3       54
4        3
5        2
Name: count, dtype: int64


Products per order


count    98666.000000
mean         1.038098
std          0.226456
min          1.000000
25%          1.000000
50%          1.000000
75%          1.000000
max          8.000000
Name: product_id, dtype: float64


Payments per order


1     96479
2      2382
3       301
4       108
5        52
6        36
7        28
8        11
9         9
10        5
11        8
12        8
13        3
14        2
15        2
19        2
21        1
22        1
26        1
29        1
Name: count, dtype: int64


Reviews per order


1    98126
2      543
3        4
Name: count, dtype: int64

# Relational Structure and Baseline Population

In [64]:
orders_without_items = orders.loc[
    ~orders["order_id"].isin(items["order_id"])
].copy()

print("Orders without item records:", len(orders_without_items))

display(
    orders_without_items["order_status"]
    .value_counts()
    .to_frame("orders")
)

Orders without item records: 775


,orders
order_status,
unavailable,603
canceled,164
created,5
invoiced,2
shipped,1


In [65]:
orders_without_items_check = orders_without_items[
    ["order_id", "order_status", "customer_id"]
].copy()

orders_without_items_check["has_customer"] = (
    orders_without_items_check["customer_id"].isin(customers["customer_id"])
)

orders_without_items_check["has_payment"] = (
    orders_without_items_check["order_id"].isin(payments["order_id"])
)

orders_without_items_check["has_review"] = (
    orders_without_items_check["order_id"].isin(reviews["order_id"])
)

# Seller attribution requires an order-item record because seller_id
# is linked to order_id through the items table.
orders_without_items_check["has_seller_attribution"] = (
    orders_without_items_check["order_id"].isin(
        items.loc[items["seller_id"].notna(), "order_id"]
    )
)

display(
    orders_without_items_check.groupby("order_status")[
        [
            "has_customer",
            "has_payment",
            "has_review",
            "has_seller_attribution",
        ]
    ]
    .agg(["sum", "mean"])
    .round(3)
)

has_customer      has_payment      has_review         \
                      sum mean         sum mean        sum   mean   
order_status                                                        
canceled              164  1.0         164  1.0        161  0.982   
created                 5  1.0           5  1.0          3  0.600   
invoiced                2  1.0           2  1.0          2  1.000   
shipped                 1  1.0           1  1.0          1  1.000   
unavailable           603  1.0         603  1.0        589  0.977   

             has_seller_attribution       
                                sum mean  
order_status                              
canceled                          0  0.0  
created                           0  0.0  
invoiced                          0  0.0  
shipped                           0  0.0  
unavailable                       0  0.0

The 775 orders without item records are primarily unavailable or canceled
orders. All have associated customer and payment records, and most also
have reviews. However, seller attribution is not available because
`seller_id` is linked to `order_id` through the order-item table.

Because the consolidated baseline uses the order-item table as its base,
these orders are naturally excluded. This is appropriate for seller-level
analysis because orders without item records cannot be attributed to a
seller.

# `Reviews` Structure

Reviews cannot be merged directly into the order-item baseline because neither
`order_id` nor `review_id` is unique in the reviews table. Before aggregating
reviews to the order level, examine the relationships between review records,
orders, and customers to determine whether repeated records represent duplicate
data or distinct review activity.

In [66]:
# Review records per order
reviews_per_order = reviews.groupby("order_id").size()
multi_review_orders = reviews_per_order[reviews_per_order > 1].index

multi_order_review_summary = (
    reviews[reviews["order_id"].isin(multi_review_orders)]
    .groupby("order_id")
    .agg(
        review_rows=("review_id", "size"),
        unique_review_ids=("review_id", "nunique"),
        unique_scores=("review_score", "nunique"),
    )
)

# Review IDs associated with multiple orders
review_id_summary = (
    reviews.groupby("review_id")
    .agg(
        review_rows=("review_id", "size"),
        unique_orders=("order_id", "nunique"),
        unique_scores=("review_score", "nunique"),
    )
)

shared_review_ids = review_id_summary[
    review_id_summary["unique_orders"] > 1
].index

shared_review_customers = (
    reviews[reviews["review_id"].isin(shared_review_ids)]
    .merge(
        orders[["order_id", "customer_id"]],
        on="order_id",
        how="left",
        validate="many_to_one",
    )
    .merge(
        customers[["customer_id", "customer_unique_id"]],
        on="customer_id",
        how="left",
        validate="many_to_one",
    )
    .groupby("review_id")["customer_unique_id"]
    .nunique()
)

print("Orders with multiple review records:", len(multi_review_orders))
print(
    "Orders with multiple review scores:",
    (multi_order_review_summary["unique_scores"] > 1).sum(),
)
print(
    "Review IDs associated with multiple orders:",
    len(shared_review_ids),
)
print(
    "Shared review IDs belonging to one unique customer:",
    (shared_review_customers == 1).sum(),
)

Orders with multiple review records: 547
Orders with multiple review scores: 202
Review IDs associated with multiple orders: 789
Shared review IDs belonging to one unique customer: 789


The review table contains two distinct forms of repeated records:

- **547 orders have multiple review records.** Each has multiple distinct
  `review_id` values, and 202 contain more than one review score. These records
  therefore contain potentially meaningful review activity rather than simple
  duplicates.
- **789 review IDs are associated with multiple orders.** In every case, the
  associated orders belong to the same `customer_unique_id`, indicating that a
  single review event can be associated with multiple orders from the same
  underlying customer.

Therefore, neither `review_id` nor `order_id` should be treated as a unique key,
and review records should not be arbitrarily deduplicated. Reviews will instead
be summarized at the order level before being merged into the order-item
baseline.

# Date Conversion and Temporal Data Quality

In [67]:
review_date_cols = [
    "review_creation_date",
    "review_answer_timestamp",
]

for col in review_date_cols:
    reviews[col] = pd.to_datetime(reviews[col])

order_date_cols = [
    "order_purchase_timestamp",
    "order_approved_at",
    "order_delivered_carrier_date",
    "order_delivered_customer_date",
    "order_estimated_delivery_date",
]

orders[order_date_cols] = orders[order_date_cols].apply(pd.to_datetime)

items["shipping_limit_date"] = pd.to_datetime(
    items["shipping_limit_date"]
)

In [68]:
orders["order_status"].value_counts()

order_status
delivered      96478
shipped         1107
canceled         625
unavailable      609
invoiced         314
processing       301
created            5
approved           2
Name: count, dtype: int64

In [69]:
timestamp_cols = [
    "order_approved_at",
    "order_delivered_carrier_date",
    "order_delivered_customer_date",
    "order_estimated_delivery_date",
]

missing_by_status = (
    orders.groupby("order_status")[timestamp_cols]
    .agg(lambda x: x.isna().mean())
    .round(3)
)

display(missing_by_status)

,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
order_status,,,,
approved,0.000,1.00,1.00,0.0
canceled,0.226,0.88,0.99,0.0
created,1.000,1.00,1.00,0.0
delivered,0.000,0.00,0.00,0.0
invoiced,0.000,1.00,1.00,0.0
processing,0.000,1.00,1.00,0.0
shipped,0.000,0.00,1.00,0.0
unavailable,0.000,1.00,1.00,0.0


In [70]:
delivered = orders.loc[
    orders["order_status"] == "delivered"
].copy()

delivered["approval_hours"] = (
    delivered["order_approved_at"]
    - delivered["order_purchase_timestamp"]
).dt.total_seconds() / 3600

delivered["seller_to_carrier_days"] = (
    delivered["order_delivered_carrier_date"]
    - delivered["order_approved_at"]
).dt.total_seconds() / 86400

delivered["carrier_to_customer_days"] = (
    delivered["order_delivered_customer_date"]
    - delivered["order_delivered_carrier_date"]
).dt.total_seconds() / 86400

delivered["total_delivery_days"] = (
    delivered["order_delivered_customer_date"]
    - delivered["order_purchase_timestamp"]
).dt.total_seconds() / 86400

display(
    delivered[
        [
            "approval_hours",
            "seller_to_carrier_days",
            "carrier_to_customer_days",
            "total_delivery_days",
        ]
    ].describe()
)

,approval_hours,seller_to_carrier_days,carrier_to_customer_days,total_delivery_days
count,96464.000000,96462.000000,96469.000000,96470.000000
mean,10.276767,2.799076,9.330297,12.558217
std,20.535419,3.537077,8.759924,9.546156
min,0.000000,-171.219005,-16.096169,0.533414
25%,0.215278,0.874363,4.100046,6.766204
50%,0.343333,1.815845,7.099769,10.217477
75%,14.514583,3.575009,12.028646,15.720182
max,741.443611,125.762569,205.190972,209.628611


In [71]:
anomalies = pd.Series({
    "approval_before_purchase": (
        delivered["order_approved_at"]
        < delivered["order_purchase_timestamp"]
    ).sum(),

    "carrier_before_approval": (
        delivered["order_delivered_carrier_date"]
        < delivered["order_approved_at"]
    ).sum(),

    "delivery_before_carrier": (
        delivered["order_delivered_customer_date"]
        < delivered["order_delivered_carrier_date"]
    ).sum(),

    "delivery_before_purchase": (
        delivered["order_delivered_customer_date"]
        < delivered["order_purchase_timestamp"]
    ).sum(),
})

display(anomalies)

approval_before_purchase       0
carrier_before_approval     1350
delivery_before_carrier       23
delivery_before_purchase       0
dtype: int64

In [72]:
anomalous_carrier = orders.loc[
    orders["order_delivered_carrier_date"] < orders["order_approved_at"],
    [
        "order_id",
        "order_status",
        "order_purchase_timestamp",
        "order_approved_at",
        "order_delivered_carrier_date",
        "order_delivered_customer_date",
    ],
].copy()

anomalous_carrier["carrier_before_approval_days"] = (
    anomalous_carrier["order_approved_at"]
    - anomalous_carrier["order_delivered_carrier_date"]
).dt.total_seconds() / 86400

display(
    anomalous_carrier["carrier_before_approval_days"].describe()
)

count    1359.000000
mean        1.031333
std         4.804491
min         0.000243
25%         0.058976
50%         0.715324
75%         1.081528
max       171.219005
Name: carrier_before_approval_days, dtype: float64

**Timestamp consistency:** A small subset of orders contains lifecycle timestamps that do not follow the expected purchase → approval → carrier → delivery sequence. Inspection indicates that these inconsistencies do not have a single cause: some records contain clearly anomalous carrier timestamps, while others contain unusually late approval timestamps despite otherwise plausible fulfillment sequences. Because the correct timestamp cannot be reliably inferred, source timestamps were retained unchanged. Derived duration features based on inconsistent timestamp pairs should be treated as invalid or missing when used in subsequent analysis.

# Aggregate One-to-Many Order Attributes

In [73]:
payment_summary = (
    payments.groupby("order_id")
    .agg(
        payment_value=("payment_value", "sum"),
        payment_installments=("payment_installments", "max"),
        payment_count=("payment_sequential", "count"),
        payment_type_count=("payment_type", "nunique"),
    )
    .reset_index()
)

primary_payment = (
    payments.loc[
        payments.groupby("order_id")["payment_value"].idxmax(),
        ["order_id", "payment_type"],
    ]
    .rename(columns={"payment_type": "primary_payment_type"})
)

payment_summary = payment_summary.merge(
    primary_payment,
    on="order_id",
    how="left",
    validate="one_to_one",
)

In [74]:
reviews_sorted = reviews.sort_values(
    ["order_id", "review_answer_timestamp"]
)

review_summary = (
    reviews_sorted.groupby("order_id")
    .agg(
        avg_review_score=("review_score", "mean"),
        last_review_score=("review_score", "last"),
        review_count=("review_id", "size"),
    )
    .reset_index()
)

In [75]:
assert payment_summary["order_id"].is_unique
assert review_summary["order_id"].is_unique

# Construct Order-Item Baseline

In [76]:
olist_df = (
    items
    .merge(
        orders,
        on="order_id",
        how="left",
        validate="many_to_one",
    )
    .merge(
        products,
        on="product_id",
        how="left",
        validate="many_to_one",
    )
    .merge(
        sellers,
        on="seller_id",
        how="left",
        validate="many_to_one",
    )
    .merge(
        customers,
        on="customer_id",
        how="left",
        validate="many_to_one",
    )
    .merge(
        payment_summary,
        on="order_id",
        how="left",
        validate="many_to_one",
    )
    .merge(
        review_summary,
        on="order_id",
        how="left",
        validate="many_to_one",
    )
)

In [77]:
assert len(olist_df) == len(items)

# Baseline Data Quality Audit

In [79]:
key_cols = [
    "order_id",
    "product_id",
    "seller_id",
    "customer_id",
    "customer_unique_id",
]

display(olist_df[key_cols].isna().sum())

order_id              0
product_id            0
seller_id             0
customer_id           0
customer_unique_id    0
dtype: int64

In [80]:
print(
    "Duplicate order-item keys:",
    olist_df.duplicated(
        subset=["order_id", "order_item_id"]
    ).sum(),
)

Duplicate order-item keys: 0


In [81]:
missing_summary = (
    olist_df.isna()
    .sum()
    .to_frame("missing_count")
)

missing_summary["missing_pct"] = (
    missing_summary["missing_count"] / len(olist_df) * 100
).round(2)

display(
    missing_summary[
        missing_summary["missing_count"] > 0
    ].sort_values("missing_pct", ascending=False)
)

,missing_count,missing_pct
order_delivered_customer_date,2454,2.18
product_category_name,1603,1.42
product_name_lenght,1603,1.42
product_description_lenght,1603,1.42
product_photos_qty,1603,1.42
order_delivered_carrier_date,1194,1.06
review_count,942,0.84
last_review_score,942,0.84
avg_review_score,942,0.84
product_width_cm,18,0.02


In [82]:
no_review_orders = (
    olist_df.loc[
        olist_df["review_count"].isna(),
        ["order_id", "order_status"],
    ]
    .drop_duplicates()
)

print("Orders without reviews:", len(no_review_orders))

display(
    no_review_orders["order_status"]
    .value_counts()
    .to_frame("orders")
)

Orders without reviews: 749


,orders
order_status,
delivered,646
shipped,75
canceled,17
processing,6
invoiced,5


In [83]:
no_payment_orders = (
    olist_df.loc[
        olist_df["payment_value"].isna(),
        ["order_id", "order_status"],
    ]
    .drop_duplicates()
)

display(no_payment_orders)

,order_id,order_status
84389,bfbd0f9bdef84302105ad712db648a6c,delivered


In [84]:
# Fill null review counts with 0
olist_df["review_count"] = (
    olist_df["review_count"]
    .fillna(0)
    .astype(int)
)

**Missing values:** Remaining missing values reflect unavailable source attributes or lifecycle events that did not occur or were not recorded. These values are retained rather than globally imputed because appropriate treatment depends on the downstream analytical use. review_count is set to 0 when no review record exists because this represents an observed absence of review activity. Model-specific missing-value treatment will be performed during feature preparation as needed.

In [88]:
print("Shape:", olist_df.shape)
print("Unique orders:", olist_df["order_id"].nunique())
print("Unique sellers:", olist_df["seller_id"].nunique())
print("Unique products:", olist_df["product_id"].nunique())

olist_df.info()

Shape: (112650, 37)
Unique orders: 98666
Unique sellers: 3095
Unique products: 32951
<class 'pandas.DataFrame'>
RangeIndex: 112650 entries, 0 to 112649
Data columns (total 37 columns):
 #   Column                         Non-Null Count   Dtype         
---  ------                         --------------   -----         
 0   order_id                       112650 non-null  str           
 1   order_item_id                  112650 non-null  int64         
 2   product_id                     112650 non-null  str           
 3   seller_id                      112650 non-null  str           
 4   shipping_limit_date            112650 non-null  datetime64[us]
 5   price                          112650 non-null  float64       
 6   freight_value                  112650 non-null  float64       
 7   customer_id                    112650 non-null  str           
 8   order_status                   112650 non-null  str           
 9   order_purchase_timestamp       112650 non-null  datetime64[us]

In [85]:
assert len(olist_df) == len(items)
assert olist_df[["order_id", "order_item_id"]].duplicated().sum() == 0

print("Final baseline shape:", olist_df.shape)
print("Unique orders:", olist_df["order_id"].nunique())

Final baseline shape: (112650, 37)
Unique orders: 98666


### Export consolidated dataset

In [86]:
olist_df.to_csv(
    "../01_data/preprocessed/olist_order_item_baseline.csv",
    index=False,
)